### ISSUE TITLE
Test data downloaded too many times in parallel pytest

### ISSUE THREAD

**Issue by cfhammill** [neuropoly/spinalcordtoolbox#2957]
### Description

Parallel pytest downloads tests numprocesses + 1 times, instead of once.

### Steps to Reproduce

I set numprocesses to 1 in setup.cfg to avoid swamping my machine then run

```
source python/etc/profile.d/conda.sh
conda activate venv_sct
pytest
```

and observe

```
Status: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6.82M/6.82M [00:02<00:00, 3.03MB/s]
Done!

=========================================================================================================== test session starts ===========================================================================================================
platform linux -- Python 3.6.12, pytest-6.1.1, py-1.9.0, pluggy-0.13.1 -- /home/chammill/documents/spinalcordtoolbox/python/envs/venv_sct/bin/python
cachedir: .pytest_cache
rootdir: /home/chammill/documents/spinalcordtoolbox, configfile: setup.cfg
plugins: console-scripts-1.0.0, cov-2.10.1, forked-1.3.0, xdist-2.1.0
[gw0] linux Python 3.6.12 cwd: /home/chammill/documents/spinalcordtoolbox
[gw0] Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56)  -- [GCC 7.3.0]
Status: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6.82M/6.82M [00:02<00:00, 3.05MB/s]
```

The two status bars for 6.82M downloads are the 2 downloads. The number of downloads increases with the number of workers. This could be fixed by either getting download caching working or by fiddling with the pytest fixtures.

A duplicate download can also be seen in the most recent travis run for 18.04 (although interestingly it didn't seem to be there for Arch linux) https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/736864114#L1040

**Comment by Drulex** [comment]
Yes this is a side-effect of using `x-dist` for parallel test distribution, each worker has its own session scope thus the initial download (session hook) is run multiple times. It appears that it is also run one additional time prior to distribution, during the initial test collection step.

I think it's preferable to keep the session hook since the test data is global to all the tests, otherwise each test will need to verify its test data before running (which was the default behavior before, but the tests didn't do this check causing issues like https://github.com/neuropoly/spinalcordtoolbox/issues/2902). 

The code in the hook is supposed to "prepare" the test data whereas right now it automatically downloads it no matter what, for lack of better implementation. Like you mentioned this could be fixed with caching, which I agree with and I saw proposed on multiple issues. I feel this is a bigger discussion since it will requiring accessing some kind of metadata to verify the integrity (whether this is published alongside the actual data, in a manifest, or somewhere else). I will create an issue to track this.

In the meantime I am proposing to add a custom option `--no-sct-download` with `action=store_true` so that one (who knows what he is doing) can choose to skip the `sct_download` step by running `pytest --no-sct-download`. This should be safe since there is a unit test that was added to verify that the tests themselves don't corrupt the test data. Just need to make sure that you start with a good test dataset. What do you think?

**Comment by jcohenadad** [comment]
Potentially useful for Pycharm users: when debugging unit tests using pytest (within Pycharm), make sure to input "-n 1" under "additional arguments". It will force single thread and download data only once.

**Comment by joshuacwnewton** [comment]
> It will force single thread and download data only once.

Just a small correction: I believe the data is downloaded twice in total with `-n 1`. We initially download the data a first time as tests are collected, then an additional `n` times depending on how `-n` is set. From https://github.com/neuropoly/spinalcordtoolbox/issues/2957#issuecomment-712139435:

> each worker has its own session scope thus the initial download (session hook) is run once per worker. It appears that it is also run one additional time prior to distribution, during the initial test collection step.

**Comment by kousu** [comment]
I'm hitting this in porting CI to Github, e.g. https://github.com/neuropoly/spinalcordtoolbox/runs/1625704027?check_suite_focus=true#step:4:2097

It's causing the tests to fail, but super intermittently. In fact I've only seen it on macOS so far (maybe the macOS scheduler is particularly prone to being unlucky with our tests?). I'm..pretty sure I've seen it at least a couple of times though, but the second is lost in the logfiles somewhere.

**Comment by kousu** [comment]
I really like using this `pytest_sessionstart()` hook though. It's very elegant. It's too bad pytest doesn't work with it; maybe there's an even more global hook we can use?

**Comment by kousu** [comment]
I might have fixed this in 585d68ca074c37d803071c3f0264a6179ddd4df6. We'll see after a few tries.

**Comment by joshuacwnewton** [comment]
@Drulex @kousu Can we close this now that #3212 has been merged?

----

One thing to mention before closing, though: I know @kousu had some other ideas about further improving how test data is downloaded and verified. They're briefly mentioned in some TODOs within #3152:

https://github.com/neuropoly/spinalcordtoolbox/blob/7329edda1c27cb2afe694b015411d5e0c5eae3c0/conftest.py#L139-L153

These TODOs also overlap with some interesting plugins I discovered via Pytest's [Sharing test data](https://docs.pytest.org/en/stable/fixture.html#sharing-test-data) section of the Fixture page.

> Another good approach is by adding the data files in the tests folder. There are also community plugins available to help managing this aspect of testing, e.g. [pytest-datadir](https://pypi.org/project/pytest-datadir/) and [pytest-datafiles](https://pypi.org/project/pytest-datafiles/).

But, I think it might be better to summarize all of this in a new issue, instead. I'd be happy to write that up. :)
